### DISCUSSION TITLE
Implement Email Events

### DISCUSSION THREAD

**Discussion by Lambels** [benbjohnson/wtf#60] [Q&A]
Hey Ben!

I can see that the event notification approach is implementing the `EventService` interface, from what I understand you've implemented this interface with `inmem.EventService` (correct me if i am wrong) and then use the `Subscription` type to encapsulate each user under his own event receiver and subscription. Then you handle each subscription under the `/events` route with `*server.handleEvents` which pushes the events to the user through a ws connection. What would be the implementation of event handling but instead of pushing the event through a ws connection, I would like to notify the user through a provided email, which would be accessed through the `userID`.

I hope that my question makes sense, if it doesn't I am happy to expand, I've been programming in go for 1 year and I am only 15 so I am not so experienced, but I am happy to learn! Also I understand if what I've said higher up is not correct or doesn't make sense so please correct me on my understanding of the program if otherwise. 😄 

Kind Regards,
Patrick A

**Comment by pegeraw** [comment]
I would implement EmailService for this as separate thing, one exception that I see is when You want to push same events through websockets and email. Then You would have to send email from handler where You use websockets using same events queue, or just push event to EventService and to EmailService in place when event happens and then they will be using different queues. EmailService doesn't need queue because it's just sending emails. You don't have to receive this event in different part of program to pass it to the user. 
English is not my main language in case when I've messed up something 😅

**Reply by Lambels** [reply]
Everything seems clear enough, thanks!

**Comment by hi019** [comment]
I'm building an RSS reader following the philosophy defined in Ben's Standard Package Layout article. My project isn't ready yet, but it's opensource and called [Doublequote](https://github.com/hi019/doublequote). I didn't implement any WS stuff, but I do have an event queue. It works like this:

* An `EventService` [interface](https://github.com/hi019/doublequote/blob/main/event.go). Through it, functions can publish and subscribe to events
* Every Event type has a struct which describes its payload, also in the file linked above
* The actual [implementation](https://github.com/hi019/doublequote/blob/main/asynq/event.go) of the service is powered by [asynq](https://github.com/hibiken/asynq)
* When a user is created, an event is [dispatched](https://github.com/hi019/doublequote/blob/d87dcb07cfad102fef0a9cdcb914a589dfc5ca2d/sql/user.go#L97)
* My event listeners (subscribers) are in a folder called listener. For example, [here](https://github.com/hi019/doublequote/blob/main/listener/user.go#L11) is the subscriber that sends a verification email to new users. The actual call to `EventService.Subscribe` occurs [here](https://github.com/hi019/doublequote/blob/main/listener/listener.go#L28)

Let me know if you have any questions

**Reply by Lambels** [reply]
First of all awesome project!

Yep that's kind of what I was looking for, I will look more into [asynq](https://github.com/hibiken/asynq) as it looks like something I will need forwards in my project!

Thx for the clear explanation!
